### PR TITLE
improve presentation of info on Query by Method Name

### LINKED ISSUE #576: [clarification]: query by method name method parameters
### Specification

Query by Method Name

### I need clarification on ...

The relationship between Query by Method Name operators and the parameters of the repository method.

### Additional information

If we're going to keep this as a required part of the spec [now that we have JDQL, `@Query`,  and `@Find` properly nailed down, I'm ready to have a probably-difficult conversation about that], then we at least need to make sure that this facility is fully-specified.

Reviewing this, it does seem to be mostly sorta OK, but:

- the allowed return types are defined only in `module-info`, in a very sketchy format, and not in the spec itself, and
- the allowed parameter types, and their relationship to the query itself are, AFICS, completely unspecified. This is less-than-trivial when we have operators like `In` and `Between`.

So there is at least some work to do here.

Also I'm also not sure I understand how to interpret shorthands like `OrderBy____(Asc|Desc)*(Asc |Desc)`. Actually that might just be an error.

### PR DESCRIPTION
**Pull request by gavinking**
Fixes #576.

### CODE REVIEW

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (njr-11) [inline spec/src/main/asciidoc/method-query.asciidoc:60]: This is probably more a question for @otaviojava , but this seems wrong. Surely these database types must be capable of at least some subset of it, otherwise how can they be supported at all in Jakarta Data?

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (njr-11) [inline spec/src/main/asciidoc/method-query.asciidoc:183]: I don't agree with getting rid of the place to document the NoSQL imitations.

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (njr-11) [inline spec/src/main/asciidoc/method-query.asciidoc:178]: Don't get rid of this. It is what allows for NoSQL implementations that are incapable of a subset of keywords, such as `Or`. The TCK relies on this for accurate testing.

**Review by njr-11 — CHANGES_REQUESTED** [review]

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (gavinking) [reply spec/src/main/asciidoc/method-query.asciidoc:60]: Well that's what we have in the table right now. Every action is marked "Not required" for these kinds of databases. It seemed strange to me too.

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (gavinking) [reply spec/src/main/asciidoc/method-query.asciidoc:183]: I think this a good change. There was only one kind of limitation here, which is much more compactly written in a single sentence than taking up that extra space in the table.

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (gavinking) [reply spec/src/main/asciidoc/method-query.asciidoc:178]: I didn't get rid of anything. I moved it.

**Comment by gavinking** [comment]
@njr-11 I recommend compiling the PDF and reading it from top to bottom, the diff is probably misleading.

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (otaviojava) [reply spec/src/main/asciidoc/method-query.asciidoc:60]: The `key-value` can only support any operation related to the key, with some exceptions. Otherwise, it will return an `UnsupportedOperationException`; the same generally happens with most wide-columns, such as Hbase, ScyllaDB, and Cassandra. 

This message was sent because all possible key operations were converted at the BasicRepository. Therefore, we decided to proceed in this direction.

⚠️ Cassandra and ScyllaDB can support if including the secondary index, but, it it not like

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (otaviojava) [reply spec/src/main/asciidoc/method-query.asciidoc:183]: We have already informed that the database that does not support it will throw an UnsupportedOperationException.

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (gavinking) [reply spec/src/main/asciidoc/method-query.asciidoc:178]: > Don't get rid of this. It is what allows for NoSQL implementations that are incapable of a subset of keywords, such as Or. 

FTR, `Or` is not on the list of things which are only supported by a subset of databases. It was listed as 

> Not required for: Key-value, Wide-Column

But so was `findBy`, so that wasn't a meaningful additional restriction.

I hope you see how the presentation in terms of tables was making this more confusing.

Comment (gavinking) [reply spec/src/main/asciidoc/method-query.asciidoc:178]: > Not required for: Key-value, Wide-Column

Indeed, this was listed for _every single keyword_.

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (njr-11) [reply spec/src/main/asciidoc/method-query.asciidoc:60]: > all possible key operations were converted at the BasicRepository. Therefore, we decided to proceed in this direction.

It's probably worth pointing out that since then a number of operations have been removed from `BasicRepository`, including some that were id-based, so that might not be true anymore.  In any case I would say it seems wrong because Jakarta Data does not force the use of `BasicRepository`.

Comment (gavinking) [reply spec/src/main/asciidoc/method-query.asciidoc:60]: OK, so then we need to sort out what are the real requirements here.

I'm not sure that such discussions should block this PR, since that's a longer conversation, which will be facilitated by having the current requirements presented in a better/clearer way.

Comment (gavinking) [reply spec/src/main/asciidoc/method-query.asciidoc:60]: OTOH, I guess I would say that if key-value databases only support key-based operations, then what is the value in having QbMN for those databases? They can already do everything possible with `@Delete` and `@Find`.

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (njr-11) [reply spec/src/main/asciidoc/method-query.asciidoc:183]: > I think this a good change. There was only one kind of limitation here, which is much more compactly written in a single sentence than taking up that extra space in the table.

I think the compactness is an inferior way to convey the information to the user that will be more difficult for them to understand, even more so as more restrictions are identified with NoSQL database types that need to be documented.

**Review by njr-11 — APPROVED** [review]
Some of the changes in this pull are nice, some I don't like and think make the spec more difficult to understand. Much of it is a question of preference. Given that so much of it is about NoSQL, I'll let Otavio make the call on these things.

**Comment by gavinking** [comment]
And I don't see how it's more difficult to understand (or easy to miss) the following admonitions:

Screenshot 2024-04-08 at 6 16 24 PM
Screenshot 2024-04-08 at 6 16 35 PM

**Review by njr-11 — CHANGES_REQUESTED** [review]
> And I don't see how it's more difficult to understand (or easy to miss) the following admonitions:

That may be so, but I just realized when look at those examples, you have also changed the requirement allowing these patterns to be rejected at compile time, which breaks the TCK, so now I've got to reverse my approval on this one.

I thought the idea here was that these sort of methods, unlike some of the others, being dependent on what a database can do, would be something to be determined at run time based on which type and version of the database is observed to be present.

**Comment by gavinking** [comment]
A Jakarta Data provider knows what kind of datastore it targets, and it knows the capabilities of that datastore. There's no good reason it can't reject a repository method at compile time if the repository method makes use of capabilities that the provider/datastore doesn't have.

> I thought the idea here was that these sort of methods, unlike some of the others, being dependent on what a database can do, would be something to be determined at run time based on which type and version of the database is observed to be present.

I don't see why you would think that.

> breaks the TCK

So we fix it. Horses go before carts.

**Comment by njr-11** [comment]
> > breaks the TCK
> 
> So we fix it. Horses go before carts.

Unfortunately, this is a type of problem that we haven't been able to find a solution for in the TCK. Failing at compile time breaks the whole TCK application. The only way we got around that last time was removing TCK coverage for a single scenario and rationalizing that it did not impact other usage because we could continue trapping for UnsupportedOperationException in the other database type/version category of places -- but that's changed under this PR.  I'm not willing to put the TCK into an unrecoverable state at this point in the release cycle.  I don't even see how the UnsupportedOperationException impacts your implementation here. This is all NoSQL stuff that would be raising it. If we find a solution next version, the spec can be compatibly updated (changes to exception paths are allowed) to allow compile time failure then. This is a case where there is no clear benefit to making the change but obvious problems with allowing it in.

**Comment by gavinking** [comment]
> I don't even see how the `UnsupportedOperationException` impacts your implementation here. 

It doesn't impact _my_ implementation at all. But @otaviojava has an annotation-processor-based implementation which targets NoSQL.

**Comment by gavinking** [comment]
Anyway, fine, I'll back out that change for now, but in future the TCK is going to need well-defined profiles for each kind of database.

**Review by njr-11 — APPROVED** [review]

**Comment by gavinking** [comment]
> I'll back out that change for now

Done by 2e66b2635e5f3ac798c07d53727ff20caf0370af.

**Comment by gavinking** [comment]
@otaviojava?

File: spec/src/main/asciidoc/method-query.asciidoc
Comment (otaviojava) [reply spec/src/main/asciidoc/method-query.asciidoc:60]: That is my opinion as well.
It does not make sense to key-value database this feature.

**Review by otaviojava — APPROVED** [review]

**Pull request merged**
